DEV: add a conda environment.yml with all development dependencies

rgommers opened the pull request:
There are enough dependencies now that it's too hard to remember. This set is comprehensive, so a single
`conda env create -f environment.yml` gives you everything in a fresh environment named `numpy-dev`.

rgommers committed:
DEV: add a conda environment.yml with all development dependencies

There are enough dependencies now that it's too hard to remember.
This set is comprehensive, so a single
`conda env create -f environment.yml` gives you everything in a fresh
environment named numpy-dev.

[ci skip]

mattip commented on environment.yml:
```diff
@@ -0,0 +1,34 @@
+  - conda-forge
+dependencies:
+  - python
+  - cython
```
```suggestion
  - cython=0.29.22
```

rgommers replied:
No, I only want to pin things that are very likely to cause problems. This is unnecessary, and also wrong in the corresponding requirements.txt file.

charris replied:
A more explicit name would help, maybe `conda-environment.yml`?

rgommers replied:
`environment.yml` is the standard name for this file in other projects and in educational material (example: [SW Carpentry](https://carpentries-incubator.github.io/introduction-to-conda-for-data-scientists/04-sharing-environments/index.html)). It's also what IDEs and tools like GitHub will look for if they gain conda support. So renaming is not a good idea.

Analogy: you wouldn't call something `pypi_requirements.txt` either.

rgommers replied:
Pandas equivalent: https://github.com/pandas-dev/pandas/blob/master/environment.yml

mattip commented on environment.yml:
```diff
@@ -0,0 +1,34 @@
+  - python
+  - cython
+  - compilers
+  - openblas
```
hmm. Shouldn't this be using our version of OpenBLAS?

rgommers replied:
No, it should use the conda-forge provided one.

mattip commented on environment.yml:
```diff
@@ -0,0 +1,34 @@
+  - scipy
+  - pandas
+  - matplotlib
+  - pydata-sphinx-theme
```
```suggestion
  - pydata-sphinx-theme-0.5.0
```

rgommers replied:
Also no, don't want to pin this.

mattip commented on environment.yml:
```diff
@@ -0,0 +1,34 @@
+  # Used in some tests
+  - cffi
+  - pytz
+  - pickle5
```
pickle5 for python<=3.7 and implementation != PyPy

rgommers replied:
Maybe better to leave it out then, it's a dont-care anyway.

rgommers replied:
I removed this, resolving.

mattip reviewed (commented):
Using conda could be mentioned in the documentation:
- for [building docs](https://numpy.org/devdocs/docs/howto_build_docs.html#instructions)
- for [building for tests and docs](https://numpy.org/devdocs/dev/index.html#test-coverage)

and maybe even describe conda in [describing virtualenvs](https://numpy.org/devdocs/dev/development_environment.html#using-virtualenvs)

How do we keep the versions synced with the requirements.txt files? Does dependabot see this file?

rgommers commented:
> and maybe even describe conda in [describing virtualenvs](https://numpy.org/devdocs/dev/development_environment.html#using-virtualenvs)

I think that's the one place I'd like to update.

> How do we keep the versions synced with the requirements.txt files? Does dependabot see this file?

Dependabot doesn't see it AFAIK. There's only a few pins, which are easy enough to update by hand if any test fails. I don't expect that to happen often, probably only once a year or less.

rgommers committed:
DOC: update development environment docs for conda virtual environments

bashtage commented:
Should this be included in CI to make sure it doesn't get out of date?

rgommers commented:
> Should this be included in CI to make sure it doesn't get out of date?

I think that's a good idea, however I'd like to file a follow-up issue for that. Adding an environment file is a useful enhancement in itself, and I don't have the bandwidth to switch some CI jobs to conda right now.

charris commented:
This could use a release note, probably under `new_feature`.  I wonder if this and the various `*_requirements.txt` files should be included in the source distribution. We include `runtests.py` and `tox.ini`.

rgommers commented:
> This could use a release note, probably under `new_feature`. I wonder if this and the various `*_requirements.txt` files should be included in the source distribution? We include `runtests.py` and `tox.ini`.

I think yes, they should. Not that they'll be used much that way (the only way to do so is to manually download the tarball and go from there), but an sdist is also a snapshot of a release with archiving/historical value. I'll update this.

rgommers committed:
REL: do a better job of including files in repo root in sdist

Not only environment.yml is added, setup.cfg was missing too. Current
result of `ls -a` in the top-level directory of the unpacked sdist:

```
.                     LICENSE.txt              pytest.ini                site.cfg.example
..                    linter_requirements.txt  README.md                 test_requirements.txt
benchmarks            MANIFEST.in              release_requirements.txt  THANKS.txt
doc                   numpy                    runtests.py               tools
doc_requirements.txt  pavement.py              setup.cfg                 tox.ini
environment.yml       PKG-INFO                 setup.py                  versioneer.py
INSTALL.rst.txt       pyproject.toml
```

rgommers commented:
I fixed the `MANIFEST.in` includes. `setup.cfg` was missing too. Changed to `include *.*` followed by excluding things, that way if we make a mistake we include a little too much which is better than including too little.

Decided against a release note. This is not a user-accessible feature from a release artifact, so it only belongs in the developer docs.

charris commented:
Thanks Ralf.
